UXD-302 Button Documentation mdx (#738)
acl-services/paprika: packages/Button/stories/Button.mdx

import { Meta, Props, Story } from "@storybook/addon-docs/blocks";
import PlusIcon from "@paprika/icon/lib/Add";
import InfoIcon from "@paprika/icon/lib/InfoCircle";
import ButtonSubmit from "./examples/ButtonSubmit";
import CloseButtonRef from "./examples/CloseButtonRef";
import Focus from "./examples/Focus";
import NewRef from "./examples/NewRef";
import OldRef from "./examples/OldRef";
import Button from "../src/";

## More Examples

> ### Index

Reviewer: This is awesome 👍

- <a href="#basic-button">Basic Button</a>

  - <a href="#kinds">Kinds</a>

  - <a href="#sizes">Sizes</a>

  - <a href="#states">States</a>

  - <a href="#submit">Submit</a>

- <a href="#close-button">Close Button</a>
- <a href="#link-button">Link Button</a>
- <a href="#icon-button">Icon Button</a>

### Basic Button

To add a basic button import the button and add the `<Button>` component to your project.

```js
<Button>Click Me!</Button>
```

##### Result

> <Button>Click Me!</Button>

Reviewer: I think this is redundant – we have a real basic Usage example above in the README. Lets go straight into the Kinds example.

#### Kinds

Differant projects require differant buttons, luckily the Button component can make us of the `kind` property to create a variety of buttons for differant use cases.

> ```js

Reviewer: I find the `jsx` syntax highlighting more pleasing 😁

<Button kind="Button.types.kind.DEFAULT">default</Button>
<Button kind="Button.types.kind.PRIMARY">primary</Button>
<Button kind="Button.types.kind.SECONDARY">secondary</Button>
<Button kind="Button.types.kind.DESTRUCTIVE">destructive</Button>
<Button kind="Button.types.kind.FLAT">flat</Button>
<Button kind="Button.types.kind.MINOR">minor</Button>
<Button kind="Button.types.kind.LINK">link</Button>
```

##### Result

<Button kind="default">default</Button>
<Button kind="primary">primary</Button>
<Button kind="secondary">secondary</Button>
<Button kind="destructive">destructive</Button>
<Button kind="flat">flat</Button>
<Button kind="minor">minor</Button>

> <Button kind="link">link</Button>

Reviewer: I think it would be nice two wrap these results with a bit of padding.  

At least a `<p>`, but you could also create a new helper in `.storybook/assets/common.styles.js` that has a bit more bottom margin to space it out from the following example.  

Or maybe just use the `<ButtonStory>` component?

#### Sizes

Buttons can be adjusteds to be a variety of sizes using the `size` proeprty or the `isFullWidth` proeprty.

```js
<Button size="Button.types.size.SMALL">small</Button>
<Button size="Button.types.size.MEDIUM">medium</Button>
<Button size="Button.types.size.LARGE">large</Button>
<Button isFullWidth>full width</Button>
```

##### Result

<Button size="small">small</Button>
<Button size="medium">medium</Button>
<Button size="large">large</Button>
<Button isFullWidth>full width</Button>

#### States

Buttons can be display differant states by adjusting making use of the `isActive`, `isPending`, and `isDisabled` properties.

```js

>     <Button isActive>active</Button>
>     <Button isDisabled>disabled</Button>
>     <Button isPending>pending</Button>

Reviewer: ```suggestion
<Button isActive>active</Button>
<Button isDisabled>disabled</Button>
<Button isPending>pending</Button>
```

```

##### Result

<Button isActive>active</Button>
<Button isDisabled>disabled</Button>
<Button isPending>pending</Button>

#### Submit

To turn a Button into a submit Button by adding the `isSubmit` property.

```js
<form
  onSubmit={event => {
    alert("submitted");
    event.preventDefault();
  }}
>

>   <input type="text" defaultValue="press enter while focus" />

Reviewer: It's so ugly 🙈  

The Paprika `<Input>` would be much nicer, but you'd have to change the `<ButtonSubmit>` story too, and add some custom CSS... 😓

  <Button isSubmit isSemantic={false}>
    submit
  </Button>
</form>
```

##### Result

<ButtonSubmit />

### Close Button

To make use of a Close Button use the `<Button.Close>` element.

```js
<p>
<Button.Close onClick={clickHandler} ref={buttonRef} />
</p>
<p>
<Small>This CloseButton will capture the focus after 1 second.</Small>
</p>
```

##### Result

<CloseButtonRef />

### Link Button

> To make use of a Link Button use the `<Link.Button>` element.

Reviewer: ```suggestion
To make use of a Link Button use the `<Button.Link>` element.
```

<Button.Link href="https://youtu.be/IdkCEioCp24?t=92" size="large">
    Link large
</Button.Link>
<Button.Link href="https://youtu.be/IdkCEioCp24?t=92" shouldOpenNewTab>
    Link in new tab
</Button.Link>
<Button.Link kind="secondary" icon={<PlusIcon />} href="https://youtu.be/IdkCEioCp24?t=92" shouldOpenNewTab>
    Button link with icon
</Button.Link>
```

##### Result

<Button.Link kind="primary" href="https://youtu.be/IdkCEioCp24?t=92">
    Link
</Button.Link>
<Button.Link kind="secondary" href="https://youtu.be/IdkCEioCp24?t=92" shouldOpenNewTab>
    Link in new tab
</Button.Link>
<Button.Link href="https://youtu.be/IdkCEioCp24?t=92" size="small">
    Link small
</Button.Link>
<Button.Link href="https://youtu.be/IdkCEioCp24?t=92">Link medium</Button.Link>
<Button.Link href="https://youtu.be/IdkCEioCp24?t=92" size="large">
    Link large
</Button.Link>
<Button.Link href="https://youtu.be/IdkCEioCp24?t=92" shouldOpenNewTab>
    Link in new tab
</Button.Link>
<Button.Link kind="secondary" icon={<PlusIcon />} href="https://youtu.be/IdkCEioCp24?t=92" shouldOpenNewTab>
    Button link with icon
</Button.Link>

### Icon Button

> The Icon Button allows users to easily make use of buttons that contain icons or use icons as buttons.

Reviewer: ```suggestion
To make use of an Icon Button use the `<Button.Icon>` element. The Icon Button allows users to easily render a square button that uses _only_ an icon as the label.

For accessibility, it is important to also provide a text label for Icon Buttons by utilizing the `a11yText` prop.
```

```js
<Button.Icon  size="small">
    <InfoIcon />
</Button.Icon>
<Button.Icon >
    <InfoIcon />
</Button.Icon>
<Button.Icon  size="large">
    <InfoIcon />
</Button.Icon>
<Button.Icon  kind="flat">
    <InfoIcon />
</Button.Icon>
<Button.Icon  kind="minor">
    <InfoIcon />
</Button.Icon>
<Button.Icon  icon="🐐" kind="secondary" size="small">
    <span role="img" aria-label="Llama">
    🦙
    </span>

> </Button.Icon>

Reviewer: Let's cut this one ✂️ 
It's sort of a silly edge case test, not how we _should_ use the component.
